File: src/potencial.py
import numpy as np
import os


def potencial(t, den, dy, dz, hb, imax, jmax, kn, aa, cc, source):
    try:
        dy2 = np.float64(np.power(dy, 2))
        dz2 = np.float64(np.power(dz, 2))
        s1 = np.float64(2.0 * (1.0 / dy2 + 1.0 / dz2))

        a = b = c = d = f = np.zeros((imax, jmax), dtype=np.float64)

        for j in range(0, jmax):
            for i in range(0, imax):
                a[i, j] = np.float64((1.0 / s1) * (1.0 / dy2 + aa[i, j] / (2.0 * dy)))
                b[i, j] = np.float64((1.0 / s1) * (1.0 / dy2 - aa[i, j] / (2.0 * dy)))
                c[i, j] = np.float64((1.0 / s1) * (1.0 / dz2 + cc[i, j] / (2.0 * dz)))
                d[i, j] = np.float64((1.0 / s1) * (1.0 / dz2 - cc[i, j] / (2.0 * dz)))
                f[i, j] = np.float64(source[i, j] * (1.0 / s1))

        print(a, b, c, d, f)
        # SOR parameters
        itera = 9001    # Relaxation Counter
        omega = np.float64(1.7)     # 1 < omega < 2
        eps1 = np.float64(1.0E-05)   # threshold

        # SOR start

        # Calculus: GRID INTERNAL POINTS
        anormf = np.float64(0.0)

        for ii in range(1, jmax - 1):
            for jj in range(1, imax - 1):
                anormf = np.float64(anormf + np.abs(f[jj, ii]))

        # SOR loop
        to = np.float64(0.0)

        for K in range(1, itera):
            anorm = np.float64(0.0)
            # print("Interação -> {}".format(K))
            for J in range(0, jmax):
                t[0, J] = t[imax - 1, J]    # periodic condition

            for kx in range(0, jmax):
                if kx == 0:
                    aux = [1, 1]
                if kx > 0 and kx < jmax - 1:
                    aux = [1, -1]
                if kx == jmax - 1:
                    aux = [-1, -1]

                for ky in range(1, imax - 1):
                    a1 = a[ky, kx] * t[ky + 1, kx]
                    print(a1)
                    a2 = b[ky, kx] * t[ky - 1, kx]
                    print(a2)
                    a3 = c[ky, kx] * t[ky, kx + aux[0]]
                    print(a3)
                    a4 = d[ky, kx] * t[ky, kx + aux[1]]
                    print(a4)
                    to = a1 + a2 + a3 + a4 - f[ky, kx]
                    print("to: {}".format(to))

                    if kx == jmax - 1:
                        resd = omega * (to - t[ky, kx])
                        anorm = anorm + np.abs(resd / omega)

                    t[ky, kx] = t[ky, kx] + omega * (to - t[ky, kx])

            print("anorm: {} eps1: {} anormf: {}".format(anorm, eps1, anormf))
            if anorm < (eps1 * anormf):
                print('Omega = ' + str(omega))
                print('Interações = ' + str(K))
                print('Epsilon = ' + str(eps1))
                error = eps1 * anorm
                print('Error = ' + str(error))
                print('')
                return t
    except Exception as e:
        print(e)

File: src/fortran/potencial.f90
C   ================================
C     Potencial Electrico:  En MKS
C   ================================
      SUBROUTINE POTENCIAL(T,DY,DZ,IMAX,JMAX,AA,CC,SOURCE)
      PARAMETER (NY=81,NZ=121,PI=3.141592654)
      DIMENSION A(NY, NZ),   B(NY, NZ),  C(NY, NZ),
     *          D(NY, NZ),   F(NY, NZ),  SOURCE(NY, NZ),
     *          T(NY, NZ),  AA(NY, NZ),  CC(NY, NZ)

C   --------------------------------------------------------------
C   Despues de la discretizacion
C   T(i,j)= A*T(i+1,j) + B*T(i-1,j) + C*T(i,j+1) + D*T(i,j-1) - F
C   --------------------------------------------------------------
Cf2py intent(in) T, DY, DZ, IMAX, JMAX, AA, CC, SOURCE
Cf2py intent(out) T
C   *****************************
C    COEFICIENTES A, B, C, D, F
C
             DY2 = DY*DY
             DZ2 = DZ*DZ
              S1 = 2.0*(1.0/DY2 + 1.0/DZ2)

      DO 40 J=1,JMAX
      DO 39 I=1,IMAX
         A(I,J) = (1.0/S1)*(1.0/DY2 + AA(I,J)/(2.0*DY))
         B(I,J) = (1.0/S1)*(1.0/DY2 - AA(I,J)/(2.0*DY))
         C(I,J) = (1.0/S1)*(1.0/DZ2 + CC(I,J)/(2.0*DZ))
         D(I,J) = (1.0/S1)*(1.0/DZ2 - CC(I,J)/(2.0*DZ))
         F(I,J) = SOURCE(I,J)*(1.0/S1)
   39 CONTINUE
   40 CONTINUE
C -----------------------------
C         PARAMETROS DE SOR
C -----------------------------
C
      ITERA = 9000        !-- Contador de relajacion
      OMEGA = 1.7         !--  1 < OMEGA < 2
      EPS1  = 1.0E-05     !-- Tolerancia
C
C    -------------------------------------
C     Inicio de SOR
C     CALCULO: PUNTOS INTERNOS DE LA MALLA
C    -------------------------------------
      ANORMF = 0.0
      DO J=2,JMAX-1
      DO I=2,IMAX-1
      ANORMF = ANORMF + ABS(F(I,J))
      ENDDO
      ENDDO
C
      DO 300 K = 1,ITERA    !--- LOOP de SOR
      ANORM = 0.0
      CHGMAX = 0.0
      DO 110 J = 1,JMAX
        T(1,J) = T(IMAX,J)  !--- Condicion periodica
C           T(1,J) = T(2,J)/2.
C        T(IMAX,J) = T(1,J)
C           T(I,1) = T(I,2)/2
C        T(I,JMAX) = T(I,JMAX-1)/2.
      DO 105 I = 2,IMAX-1
        IF(J.EQ.1) GO TO 85
        IF(J.GT.1.AND.J.LT.JMAX) GO TO 86
        IF(J.EQ.JMAX) GO TO 87
C--------o
   85   BOTT = T(I,J+1)              !-- Bottom boundary dT/dZ=0
          TO = A(I,J)*T(I+1,J)+B(I,J)*T(I-1,J)+C(I,J)*T(I,J+1)
     *          +D(I,J)*BOTT-F(I,J)
      GO TO 99
C--------o
   86     TO =  A(I,J)*T(I+1,J)+B(I,J)*T(I-1,J)+C(I,J)*T(I,J+1)
     *          +D(I,J)*T(I,J-1)-F(I,J)
      GO TO 99
C--------o
   87    TOP = T(I,J-1)              !-- Top boundary  dT/dZ=0
          TO =  A(I,J)*T(I+1,J)+B(I,J)*T(I-1,J)+C(I,J)*TOP
     *          +D(I,J)*T(I,J-1)-F(I,J)
C--------o
          RESD = OMEGA*( TO-T(I,J) )
          ANORM = ANORM+ABS(RESD/OMEGA)
          IF(CHGMAX.LT.EPS1) CHGMAX = ABS(RESD)
   99     T(I,J) = T(I,J) + OMEGA*( TO-T(I,J) )
  105 CONTINUE
  110 CONTINUE
      IF(ANORM.LT.EPS1*ANORMF) GO TO 335
  300 CONTINUE                       !-- FIN DE ITERA
      WRITE(*,*)(' NOT CONVERGED!')
      STOP
  335 CONTINUE
      WRITE(*,550)OMEGA
  550 FORMAT('Omega:',F12.7)
      WRITE(*,600)K
  600 FORMAT('Interações:',I6)
      WRITE(*,610)EPS1
  610 FORMAT('Epsilon:', 1PE12.4)
      ERROR = EPS1*ANORM
      WRITE(*,620)ERROR
  620 FORMAT('Error:', 1PE12.4)
      WRITE(*,*)' '
      END
C
C